ansible_builder/main.py

```diff
@@ -145,62 +147,63 @@ def get_dependency(self, entry):
-            if not dep_file:
-                continue
-            namespace, name = CD.namespace_name()
-            ret.append(os.path.join(namespace, name, dep_file))
-        return ret
-
 
 class Containerfile:
     newline_char = '\n'
 
     def __init__(self, definition,
                  filename=constants.default_file,
                  build_context=constants.default_build_context,
-                 base_image=constants.default_base_image):
+                 base_image=constants.default_base_image,
+                 container_runtime=constants.default_container_runtime,
+                 tag=constants.default_tag):
 
         self.build_context = build_context
         os.makedirs(self.build_context, exist_ok=True)
         self.definition = definition
         self.path = os.path.join(self.build_context, filename)
         self.base_image = base_image
+        self.container_runtime = container_runtime
+        self.tag = tag
+        self.steps = []
 
-    def build_steps(self):
+    def prepare_galaxy_steps(self):
         self.steps = [
             "FROM {}".format(self.base_image),
             ""
         ]
-        requirements_path = self.definition.get_dependency('galaxy')
-        if requirements_path:
+        galaxy_requirements_path = self.definition.get_dependency('galaxy')
+        if galaxy_requirements_path:
+            # name is most likely "requirements.yml"
+            galaxy_requirements_name = os.path.basename(galaxy_requirements_path)
             # TODO: what if build context file exists? https://github.com/ansible/ansible-builder/issues/20
-            shutil.copy(requirements_path, self.build_context)
-            self.steps.extend(
-                GalaxySteps(
-                    os.path.basename(requirements_path)  # probably "requirements.yml"
-                )
-            )
+            dest = os.path.join(self.build_context, galaxy_requirements_name)
+            exists = os.path.exists(dest)
+            if not exists or not filecmp.cmp(galaxy_requirements_path, dest, shallow=False):
+                shutil.copy(galaxy_requirements_path, dest)
+
+            self.steps.extend(GalaxySteps(galaxy_requirements_name))
 
-        # There probably needs to be an intermidiate step here
-        # where we introspect the results of running the "galaxy steps"
-        # inside of the base image.
+    def prepare_pip_steps(self):
         python_req_path = self.definition.get_dependency('python')
         if python_req_path:
             shutil.copy(python_req_path, self.build_context)
+
+
+        command = [
+            self.container_runtime, "run", "--rm", self.tag,
+            "python", "-c", "from ansible_builder.utils import introspect; introspect()"
+        ]
+        rc, output = run_command(command, capture_output=True)
+        if rc is not 0:
+            print('No collections requirements file found, skipping ansible-galaxy install...')
+            return
```

Comment:
This should address the traceback issue filed in https://github.com/ansible/ansible-builder/issues/26, @kdelee

Reply:
This won't work, it regresses the case where there is a user-specified python file

```yaml
---
version: 1
dependencies:
  python: project/requirements.txt
```

The requirements file gets copied into the build context, but then it never gets added in the Containerfile.